modules/system/classes/extensions/PluginManager.php

```diff
@@ -120,8 +120,389 @@ protected function init(): void
+        }
+
+        foreach ($plugins as $code => $plugin) {
+            $pluginName = Lang::get($plugin->pluginDetails()['name']);
+
+            // Plugin record will be null if trying to update before install
+            try {
+                $pluginRecord = $this->getPluginRecord($plugin);
+            } catch (\Throwable $e) {
+                $pluginRecord = null;
+            }
+
+            if (!$migrationsOnly) {
+                if (
+                    !$pluginRecord?->is_frozen
+                    && ($composerPackage = $plugin->getComposerPackageName())
+                    && Composer::updateAvailable($composerPackage)
+                ) {
+                    $this->renderComponent(Info::class, sprintf(
+                        'Performing composer update for %s (%s) plugin...',
+                        $pluginName,
+                        $code
+                    ));
+
+                    Preserver::instance()->store($plugin);
+                    // @TODO: Make this not dry run
+                    $update = Composer::update(dryRun: true, package: $composerPackage);
+
+                    ($versions = $update->getUpgraded()[$composerPackage] ?? null)
+                        ? $this->renderComponent(Info::class, sprintf(
+                            'Updated plugin %s (%s) from v%s => v%s',
+                            $pluginName,
+                            $code,
+                            $versions[0],
+                            $versions[1]
+                        ))
+                        : $this->renderComponent(Error::class, sprintf(
+                            'Failed to update plugin %s (%s)',
+                            $pluginName,
+                            $code
+                        ));
+                } elseif (false /* Detect if market */) {
+                    Preserver::instance()->store($plugin);
+                    // @TODO: Update files from market
+                }
+            }
+
+            $this->renderComponent(Info::class, sprintf('Migrating %s (%s) plugin...', $pluginName, $code));
+            $this->versionManager->updatePlugin($plugin);
+
+            // Ensure any active aliases have their history migrated for replacing plugins
+            $this->migratePluginReplacements();
+        }
+
+        return true;
+    }
+
+    /*
+     * Replace plugins
+     */
+    public function migratePluginReplacements(): static
+    {
+        foreach ($this->list() as $code => $plugin) {
+            if (!($replaces = $plugin->getReplaces())) {
+                continue;
+            }
+            // @TODO: add full support for plugins replacing multiple plugins
+            if (count($replaces) > 1) {
+                throw new ApplicationException(Lang::get('system::lang.plugins.replace.multi_install_error'));
+            }
+            foreach ($replaces as $replace) {
+                $this->versionManager()->replacePlugin($plugin, $replace);
```

Comment:
Why is there a method for accessing the property?

Reply:
Public method, private property. Property should be used internally tho, this code was moved from `UpdateManager` and this refactor was overlooked. See: https://github.com/wintercms/winter/pull/1259/commits/c8bbe35b90f157ca5f4e832bf29396bf5d8d177e